File: archive/root_legacy/wnode-sdk-go/mesh/workflow.go
package mesh

import (
	"encoding/json"
	"sync"

	"github.com/wnodeltd/wnode/wnode-sdk-go"
)

type MeshWorkflowCoordinator struct {
	assignments map[string]WorkflowStepAssignment
	mu          sync.RWMutex
}

func NewMeshWorkflowCoordinator() *MeshWorkflowCoordinator {
	return &MeshWorkflowCoordinator{
		assignments: make(map[string]WorkflowStepAssignment),
	}
}

func (c *MeshWorkflowCoordinator) AssignStep(assignment WorkflowStepAssignment) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.assignments[assignment.StepID] = assignment
}

func (c *MeshWorkflowCoordinator) GetAssignments() []WorkflowStepAssignment {
	c.mu.RLock()
	defer c.mu.RUnlock()

	var list []WorkflowStepAssignment
	for _, a := range c.assignments {
		list = append(list, a)
	}
	return list
}

type MeshWorkflowWorker struct {
	client      *sdk.WnodeClient
	localNodeID string
}

func NewMeshWorkflowWorker(client *sdk.WnodeClient, localNodeID string) *MeshWorkflowWorker {
	return &MeshWorkflowWorker{
		client:      client,
		localNodeID: localNodeID,
	}
}

func (w *MeshWorkflowWorker) ExecuteStep(assignment WorkflowStepAssignment) (*WorkflowStepResult, error) {
	if assignment.NodeID != w.localNodeID {
		return nil, &sdk.WnodeWorkflowError{
			Code: "STEP_NOT_ASSIGNED_TO_NODE",
			Context: map[string]any{
				"expected": w.localNodeID,
				"received": assignment.NodeID,
			},
		}
	}

	// Reconstruct the workflow JSON
	paramsMap, _ := assignment.Params.(map[string]any)
	if paramsMap == nil {
		paramsMap = make(map[string]any)
	}
	paramsMap["blockTag"] = assignment.BlockTag

	workflowObj := map[string]any{
		"version": "1.0",
		"steps": []map[string]any{
			{
				"id":     assignment.StepID,
				"action": assignment.Action,
				"params": paramsMap,
			},
		},
	}

	workflowBytes, _ := json.Marshal(workflowObj)

	res, err := w.client.ExecuteWorkflow(sdk.ExecuteWorkflowParams{
		Workflow: string(workflowBytes),
		Params:   make(map[string]any),
	})

	if err != nil {
		return nil, err
	}

	stepHash := ""
	if len(res.Proof.StepHashes) > 0 {
		stepHash = res.Proof.StepHashes[0]
	}

	return &WorkflowStepResult{
		WorkflowID: assignment.WorkflowID,
		StepID:     assignment.StepID,
		NodeID:     w.localNodeID,
		StepHash:   stepHash,
		LocalProof: res.Proof,
	}, nil
}

File: packages/wnode-sdk-ts/src/mesh/workflow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MeshWorkflowWorker = exports.MeshWorkflowCoordinator = void 0;
const errors_1 = require("../errors");
class MeshWorkflowCoordinator {
    assignments = new Map();
    /**
     * Assigns a deterministic step to a worker node.
     */
    assignStep(assignment) {
        this.assignments.set(assignment.stepId, assignment);
    }
    getAssignments() {
        return Array.from(this.assignments.values());
    }
}
exports.MeshWorkflowCoordinator = MeshWorkflowCoordinator;
class MeshWorkflowWorker {
    client;
    localNodeId;
    integrationRegistry;
    constructor(client, localNodeId, registry) {
        this.client = client;
        this.localNodeId = localNodeId;
        this.integrationRegistry = registry;
    }
    /**
     * Deterministically executes an assigned step and produces a local ProofOfCompute.
     */
    async executeStep(assignment) {
        if (assignment.nodeId !== this.localNodeId) {
            throw new errors_1.WnodeWorkflowError('STEP_NOT_ASSIGNED_TO_NODE', {
                expected: this.localNodeId,
                received: assignment.nodeId
            });
        }
        // Execute through WnodeClient directly for strict determinism hooks
        const res = await this.client.executeWorkflow({
            workflow: JSON.stringify({
                version: "1.0",
                steps: [
                    {
                        id: assignment.stepId,
                        action: assignment.action,
                        params: { ...assignment.params, blockTag: assignment.blockTag }
                    }
                ]
            }),
            params: {}
        });
        if (!res.proof) {
            throw new errors_1.WnodeWorkflowError('PROOF_GENERATION_FAILED', {
                stepId: assignment.stepId,
                reason: 'Workflow execution did not produce a local proof'
            });
        }
        const stepHash = res.proof.stepHashes[0];
        let integrationPayloadHash;
        let integrationIntegrityProof;
        if (assignment.integrationName && this.integrationRegistry) {
            const adapter = this.integrationRegistry.getIntegration(assignment.integrationName);
            const op = assignment.integrationOperation || 'fetch';
            const res = await adapter[op](assignment.params);
            if (res.errorCode) {
                throw new errors_1.WnodeWorkflowError('INTEGRITY_REJECTED', {
                    stepId: assignment.stepId,
                    reason: `Integration error: ${res.errorCode}`
                });
            }
            integrationPayloadHash = res.payloadHash;
            integrationIntegrityProof = res.integrityProof;
        }
        return {
            workflowId: assignment.workflowId,
            stepId: assignment.stepId,
            nodeId: this.localNodeId,
            stepHash,
            localProof: res.proof,
            integrationPayloadHash,
            integrationIntegrityProof
        };
    }
}
exports.MeshWorkflowWorker = MeshWorkflowWorker;
